gensim/models/word2vec.py

```diff
@@ -1204,19 +1211,45 @@ def prepare_vocab(self, hs, negative, wv, update=False, keep_raw_vocab=False, tr
         sample = sample or self.sample
         drop_total = drop_unique = 0
 
+        # if max_vocab is specified instead of min_count
+        # pick a min_count which satisfies max_vocab as well as possible
+        if self.max_vocab is not None:
+            sorted_vocab = sorted(self.raw_vocab.keys(), key=lambda word: self.raw_vocab[word], reverse=True)
+
+            if self.max_vocab < len(sorted_vocab):
+                calc_min_count = self.raw_vocab[sorted_vocab[self.max_vocab]] + 1
+            else:
+                calc_min_count = 1
+
+            if calc_min_count > min_count:
+                logger.info("effective_min_count was set to %d due to max_vocab being set to %d" %
+                           (calc_min_count, self.max_vocab))
+                effective_min_count = calc_min_count
+            else:
+                logger.info("""specified min_count = %d is larger that min_count calculated
+                               by max_vocab = %d, using specified min_count""" % (min_count, calc_min_count))
+                effective_min_count = None
+        else:
+            # if max_vocab is not specified
+            # do not use the effective_min_count
+            effective_min_count = None
+
         if not update:
             logger.info("Loading a fresh vocabulary")
             retain_total, retain_words = 0, []
             # Discard words less-frequent than min_count
             if not dry_run:
                 wv.index2word = []
                 # make stored settings match these applied settings
-                self.min_count = min_count
+                if effective_min_count is not None:
+                    self.min_count = effective_min_count
```

Comment:
By clobbering the user-provided `min_count`, this could make it harder for later examinations of the model to understand whether the effective minimum was specified, or forced-by-the-`max_vocab`.

Reply:
If I don't set `self.min_count` to `effective_min_count` I will have to create an if branch in all further references to min_count.

Can you suggest a way around this?

Reply:
Yes - always use `effective_min_count`. It's the 'effective' value that is determined once, then used everywhere. The `min_count` was just a user-specified parameter that might or might not be promoted to  'effective' depending on a 1-time comparison to the max-vocab-derived min-count.